server/rpc.go

```diff
@@ -50,6 +51,64 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	broadcastSessMap  map[string]*BroadcastSession
+	sessLock          *sync.Mutex
+}
+
+func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	// minOrchs is maximum number of inflight sessions
+	minOrchs := int(HTTPTimeout / SegLen)
+	if len(bsm.broadcastSessions) < minOrchs {
+		bsm.checkStatus()
+	}
+
+	if len(bsm.broadcastSessions) <= 0 {
+		return nil
+	}
+
+	last := len(bsm.broadcastSessions) - 1
+	sess, sessions := bsm.broadcastSessions[last], bsm.broadcastSessions[:last]
+	bsm.broadcastSessions = sessions
+	return sess
+}
+
+func (bsm *BroadcastSessionsManager) removeSession(session *BroadcastSession) {
+	bsm.sessLock.Lock()
+	delete(bsm.broadcastSessMap, session.OrchestratorInfo.Transcoder)
+	bsm.sessLock.Unlock()
+}
+
+func (bsm *BroadcastSessionsManager) completeSession(sess *BroadcastSession) {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	if _, ok := bsm.broadcastSessMap[sess.OrchestratorInfo.Transcoder]; ok {
+		bsm.broadcastSessions = append(bsm.broadcastSessions, sess)
+	}
+}
+
+func (bsm *BroadcastSessionsManager) checkStatus() {
+	newBroadcastSessions := []*BroadcastSession{}
+
+	// include all O's being used successfully in refreshed list
+	if len(newBroadcastSessions) > 0 {
```

Comment:
I don't think this length check is really needed anymore. If we keep it, the check should exit early rather than introduce in another level of indentation, eg

```golang
if len(newBroadcastSessions) <= 0 {
  return
}
uniqueSessions := ...
```

Reply:
I placed that check in there in case `newBroadcastSessions := []*BroadcastSession{}` which will become `newBroadcastSessions, err := selectOrchestrator(node, pl)` comes back empty, so as to not go through the following loop. I'll have it exit early and not intro more indentation! Is less indentation generally seen as more "golang-y"?

Reply:
If `len(newBroadcastSessions) == 0` then the loop wouldn't have anything to execute through, but it doesn't hurt to exit early here. See here for an example: https://play.golang.org/p/5JbDbGAFkjp

> Is less indentation generally seen as more "golang-y"?

It's generally better to exit early if possible; it makes the flow more "straight through" more and easier to follow. Deep nesting is also a bit of a code smell.